From 113273a924f3c5e07f5e512a2439c70a62899756 Mon Sep 17 00:00:00 2001
From: Andy Yan <andy.yan@rock-chips.com>
Date: Fri, 13 Nov 2020 18:09:46 +0800
Subject: [PATCH] drm/rockchip: vop2: No add fb offset for afbc

When we want display from a offset of the afbc
framebuffer, afbc_hdr_hdr still need to set
from the start address of the framebufferr.
And the x/y offset set to PIC_OFFSET register.
This is different with non-afbc window.

Change-Id: I501c2071c9569dc3801f9bc588e7517d912a7972
Signed-off-by: Andy Yan <andy.yan@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index 50a01a6f31f7..0de953ae9cb9 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -1447,15 +1447,18 @@ static int vop2_plane_atomic_check(struct drm_plane *plane, struct drm_plane_sta
 	vpstate->offset = offset + fb->offsets[0];
 
 	/*
-	 * DRM_MODE_REFLECT_Y by afbc don't need to add the offset.
+	 * AFBC HDR_PTR must set to the zero offset of the framebuffer.
 	 */
-	if (vpstate->ymirror_en && !vpstate->afbc_en)
+	if (vpstate->afbc_en)
+		offset = 0;
+	else if (vpstate->ymirror_en)
 		offset += ((src->y2 >> 16) - 1) * fb->pitches[0];
 	else
 		offset += (src->y1 >> 16) * fb->pitches[0];
 
 	dma_addr = rockchip_fb_get_dma_addr(fb, 0);
 	kvaddr = rockchip_fb_get_kvaddr(fb, 0);
+
 	vpstate->yrgb_mst = dma_addr + offset + fb->offsets[0];
 	vpstate->yrgb_kvaddr = kvaddr + offset + fb->offsets[0];
 	if (fb->format->is_yuv) {
@@ -1586,6 +1589,8 @@ static void vop2_plane_atomic_update(struct drm_plane *plane, struct drm_plane_s
 		VOP_AFBC_SET(vop2, win, half_block_en, afbc_half_block_en);
 		VOP_AFBC_SET(vop2, win, hdr_ptr, vpstate->yrgb_mst);
 		VOP_AFBC_SET(vop2, win, pic_size, act_info);
+		VOP_AFBC_SET(vop2, win, pic_offset, ((src->x1 >> 16) | src->y1));
+		VOP_AFBC_SET(vop2, win, dsp_offset, (dest->x1 | (dest->y1 << 16)));
 		VOP_AFBC_SET(vop2, win, pic_vir_width, stride);
 		VOP_AFBC_SET(vop2, win, tile_num, afbc_tile_num);
 		VOP_AFBC_SET(vop2, win, xmirror, vpstate->xmirror_en);
-- 
2.35.3

